x/ibc/core/04-channel/types/packet.go

```diff
@@ -1,28 +1,42 @@
 package types
 
 import (
-	"github.com/tendermint/tendermint/crypto/tmhash"
+	"crypto/sha256"
 
+	"github.com/cosmos/cosmos-sdk/codec"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 	clienttypes "github.com/cosmos/cosmos-sdk/x/ibc/core/02-client/types"
 	host "github.com/cosmos/cosmos-sdk/x/ibc/core/24-host"
 	"github.com/cosmos/cosmos-sdk/x/ibc/core/exported"
 )
 
-// CommitPacket returns a packet commitment bytes. The commitment consists of:
-// hash(timeout_timestamp + timeout_version_number + timeout_version_height + data) from a given packet.
-func CommitPacket(packet exported.PacketI) []byte {
+// CommitPacket returns a packet commitment bytes and an error. The commitment consists of:
+// sha256_hash(timeout_timestamp + proto_marshal(timeout_height) + data) from a given packet.
+// The function will fail if the timeout height cannot be marshaled to bytes.
+func CommitPacket(cdc codec.BinaryMarshaler, packet exported.PacketI) ([]byte, error) {
+	timeoutHeight := packet.GetTimeoutHeight()
+	height, ok := timeoutHeight.(clienttypes.Height)
+	if !ok {
+		return nil, sdkerrors.Wrap(ErrInvalidPacket, "could not cast height to clienttypes.Height")
+	}
+
+	timeoutBz, err := cdc.MarshalBinaryBare(&height)
+	if err != nil {
+		return nil, err
+	}
+
 	buf := sdk.Uint64ToBigEndian(packet.GetTimeoutTimestamp())
-	buf = append(buf, sdk.Uint64ToBigEndian(packet.GetTimeoutHeight().GetVersionNumber())...)
-	buf = append(buf, sdk.Uint64ToBigEndian(packet.GetTimeoutHeight().GetVersionHeight())...)
+	buf = append(buf, timeoutBz...)
```

Comment:
Just wondering, are we worried about potential collisions here? Would it be possible to craft packet data to overlap commitments with some other timestamp, or is that exceptionally unlikely?

Reply:
Yeah, I think it could be possible. How about also including the port and channel id into the commitment?

Reply:
> Just wondering, are we worried about potential collisions here? Would it be possible to craft packet data to overlap commitments with some other timestamp, or is that exceptionally unlikely?

sounds possible to me. Lets just separate each item with a const prefix like we do with the keys:

hash("timeout-timestamp" + timeout timestamp bytes + "timeout height" + timeout height bytes + "packet data" + packet data bytes)

Reply:
but this is something that would need to be changed on the spec. Maybe there is another approach?

Reply:
I'm pretty sure `sdk.Uint64ToBigEndian` sets the uint into a pre set length byte array. So maybe it isn't possible?

Reply:
Looked into this some more. [`sdk.Uint64ToBigEndian`](https://github.com/cosmos/cosmos-sdk/blob/master/types/utils.go#L53) is not an issue. The problem is the combination of the marhalling of the height and the data bytes.

I think just changing the ordering to heightBz + timestamp + data, could still be problematic if you can somehow us the heightBz with the timestamp and then use the start of the data to represent the timestamp.

I think we can either separate by string consts as I proposed above or we can set the revision number and revision height each as using `sdk.Uint64ToBigEndian`. That way the it'd be 8 bytes timestamp + 8 bytes revision number + 8 bytes revision height + data bytes.

Thoughts @cwgoes ? This would require a spec change

Reply:
Doing the latter might be the safest and most straightforward to avoid issues with proto

Reply:
Making a spec change for this isn't a big deal, it's a problem that needs to be fixed in the spec as well.

What about 8 bytes timestamp, 8 bytes revision number, 8 bytes revision height, and the _hash_ of the data bytes? That way the preimage is constant-length and any sort of collision would require a probably infeasible search of possible packet datas.

Reply:
Doing a hash of the data bytes sounds good to me, might not be necessary, but would be safe and probably would lower any chance of attacks for issues we haven't considered

Reply:
agreed with fixed-length preimage